core/src/com/google/inject/internal/aop/AbstractGlueGenerator.java

```diff
@@ -0,0 +1,214 @@
+   * objects, and invokes the appropriate constructor/method returning the result as an object.
+   */
+  protected static final String TRAMPOLINE_DESCRIPTOR =
+      "(ILjava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;";
+
+  protected final Class<?> hostClass;
+
+  protected final String hostName;
+
+  protected final String proxyName;
+
+  private static final AtomicInteger COUNTER = new AtomicInteger();
+
+  protected AbstractGlueGenerator(Class<?> hostClass, String marker) {
+    this.hostClass = hostClass;
+    this.hostName = Type.getInternalName(hostClass);
+    this.proxyName = proxyName(hostName, marker, hashCode());
+  }
+
+  /** Generates a unique name based on the original class name and marker. */
+  private static String proxyName(String hostName, String marker, int hash) {
+    int id = ((hash & 0x000FFFFF) | (COUNTER.getAndIncrement() << 20));
+    String proxyName = hostName + marker + id;
+    if (proxyName.startsWith("java/") && !ClassDefining.hasPackageAccess()) {
+      proxyName = '$' + proxyName; // can't define java.* glue in same package
+    }
+    return proxyName;
+  }
+
+  /** Generates the enhancer/fast-class and returns a mapping from signature to invoker. */
+  public final Function<String, BiFunction> glue(SortedMap<String, Executable> glueMap) {
+    final MethodHandle invokerTable;
+    try {
+      byte[] bytecode = generateGlue(glueMap.values());
+      Class<?> glueClass = ClassDefining.define(hostClass, bytecode);
+      invokerTable = lookupInvokerTable(glueClass);
+    } catch (Throwable e) {
+      throw new GlueException("Problem generating " + proxyName, e);
+    }
+
+    // build optimized index for these signatures and bind it to the generated invokers
+    ToIntFunction<String> signatureTable = ImmutableStringTrie.buildTrie(glueMap.keySet());
+    return bindSignaturesToInvokers(signatureTable, invokerTable);
+  }
+
+  /** Generates enhancer/fast-class bytecode for the given constructors/methods. */
+  protected abstract byte[] generateGlue(Collection<Executable> members);
+
+  /** Lookup the invoker table; this may be represented by a function or a trampoline. */
+  protected abstract MethodHandle lookupInvokerTable(Class<?> glueClass) throws Throwable;
+
+  /** Combines the signature and invoker tables into a mapping from signature to invoker. */
+  private static Function<String, BiFunction> bindSignaturesToInvokers(
+      ToIntFunction<String> signatureTable, MethodHandle invokerTable) {
+
+    // single-argument method; the table must be a function from integer index to invoker function
+    if (invokerTable.type().parameterCount() == 1) {
+      return signature -> {
+        try {
+          // pass this signature's index into the table function to retrieve the invoker
+          return (BiFunction) invokerTable.invokeExact(signatureTable.applyAsInt(signature));
+        } catch (Throwable e) {
```

Comment:
should WrongMethodTypeException be handled separately here?  javadoc describes that as thrown when "if the target's type is not identical with the caller's symbolic type descriptor", as opposed to every other kind of exception, which is just propagated from the underlying call.

same below

Reply:
I don't believe we need to handle `WrongMethodTypeException` separately here.